### PR TITLE
[WIP] Wayfire package.

### PR DESCRIPTION
**Pull request by ishan9299**
###### Motivation for this change
It adds a wayfire package. The nixpkgs-wayland overlay only adds the compositor so it needs some more packages like.
- wf-shell.
- wayfire-plugins-extra.
- wcm.
To provide a desktop experience with wayfire. 
We can also add modules for wayfire.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Right now icons are missing but everything else seems to run fine. ~~Also I used the lastest commits because the wcm package needed the wf-config submodule to be at 0.6 but the latest release had 0.5 and the build was failing.~~ (Using the stable release now). 

~~The build for wayfire-plugins-extra build is failing with this warning~~ (Not including the wayfire-plugins-extra)

```
glibPreInstallPhase
installing
install flags: install
[0/1] Installing files. ersion: 0.56.0
Installing subprojects/dbus/libdbus_interface.so to /nix/store/1s0krsi0bpv42xk4w04851rmz6shnwfh-wayfire-8aa8260972edaba6d237e1a61217a4d2303edc3e/lib/wayfire
Traceback (most recent call last):
  File "/nix/store/p5v5i9zyaysw9ncvsrqjks92j2j4bcly-meson-0.56.0/lib/python3.8/site-packages/mesonbuild/mesonmain.py", line 140, in run
    return options.run_func(options)
  File "/nix/store/p5v5i9zyaysw9ncvsrqjks92j2j4bcly-meson-0.56.0/lib/python3.8/site-packages/mesonbuild/minstall.py", line 554, in run
    installer.do_install(datafilename)
  File "/nix/store/p5v5i9zyaysw9ncvsrqjks92j2j4bcly-meson-0.56.0/lib/python3.8/site-packages/mesonbuild/minstall.py", line 372, in do_install
    self.install_targets(d)
  File "/nix/store/p5v5i9zyaysw9ncvsrqjks92j2j4bcly-meson-0.56.0/lib/python3.8/site-packages/mesonbuild/minstall.py", line 482, in install_targets
    file_copied = self.do_copyfile(fname, outname, makedirs=(d.dirmaker, outdir))
  File "/nix/store/p5v5i9zyaysw9ncvsrqjks92j2j4bcly-meson-0.56.0/lib/python3.8/site-packages/mesonbuild/minstall.py", line 278, in do_copyfile
    shutil.copy2(from_file, to_file)
  File "/nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/python3.8/shutil.py", line 432, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/python3.8/shutil.py", line 261, in copyfile
    with open(src, 'rb') as fsrc, open(dst, 'wb') as fdst:
PermissionError: [Errno 13] Permission denied: '/nix/store/1s0krsi0bpv42xk4w04851rmz6shnwfh-wayfire-8aa8260972edaba6d237e1a61217a4d2303edc3e/lib/wayfire/libdbus_interface.so'
FAILED: meson-install 
/nix/store/p5v5i9zyaysw9ncvsrqjks92j2j4bcly-meson-0.56.0/bin/meson install --no-rebuild
ninja: build stopped: subcommand failed.
error: --- Error --------------------------------------------------------------------------------------------------------------------------------------------------------------------- nix-build
builder for '/nix/store/246s2knhdwlkyq13mhwm11xkbnw3iplm-wayfire-plugins-extra-b2036fc4e5935bf459b1779b89b59a73be92da9f.drv' failed with exit code 1; last 10 log lines:
    File "/nix/store/p5v5i9zyaysw9ncvsrqjks92j2j4bcly-meson-0.56.0/lib/python3.8/site-packages/mesonbuild/minstall.py", line 278, in do_copyfile
      shutil.copy2(from_file, to_file)
    File "/nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/python3.8/shutil.py", line 432, in copy2
      copyfile(src, dst, follow_symlinks=follow_symlinks)
    File "/nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/python3.8/shutil.py", line 261, in copyfile
      with open(src, 'rb') as fsrc, open(dst, 'wb') as fdst:
  PermissionError: [Errno 13] Permission denied: '/nix/store/1s0krsi0bpv42xk4w04851rmz6shnwfh-wayfire-8aa8260972edaba6d237e1a61217a4d2303edc3e/lib/wayfire/libdbus_interface.so'
  FAILED: meson-install 
  /nix/store/p5v5i9zyaysw9ncvsrqjks92j2j4bcly-meson-0.56.0/bin/meson install --no-rebuild
  ninja: build stopped: subcommand failed.

```

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -10106,4 +10106,9 @@
     github = "zupo";
     githubId = 311580;
   };
+  ishan = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:10109]: Please put this into a separate commit with the message ``maintainers: add ishan`` and please sort it alphabetically.

File: pkgs/applications/window-managers/wayfire/default.nix [hunk 2]
```diff
@@ -0,0 +1,52 @@
+{ stdenv, fetchFromGitHub
+, meson, pkgconfig, ninja
+, wayland, wayland-protocols
+, cairo, glm
+, libevdev, freetype, libinput
+, pixman, libxkbcommon, libdrm
+, libjpeg, libpng
+, libGL, mesa
+, libcap, xcbutilerrors, xcbutilwm, libxml2
+, buildDocs ? true
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/default.nix:10]:
```suggestion
```

File: pkgs/applications/window-managers/wayfire/default.nix [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+stdenv.mkDerivation rec {
+  pname = "wayfire";
+  version = "8aa8260972edaba6d237e1a61217a4d2303edc3e";
+
+  src = fetchFromGitHub {
+    owner = "WayfireWM";
+    repo = "wayfire";
+    rev = version;
+    sha256 = "sha256-r9k8e0LCDxr8tEZvcBbkvZ2RqcKXYX+/fxalKoqT15U=";
+    fetchSubmodules = true;
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/default.nix:22]: Can we package the submodules as different packages?

File: pkgs/applications/window-managers/wayfire/default.nix [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  };
+
+  nativeBuildInputs = [ pkgconfig meson ninja ];
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/default.nix:25]:
```suggestion
  nativeBuildInputs = [ pkg-config meson ninja ];
```

File: pkgs/applications/window-managers/wayfire/default.nix [hunk 5]
```diff
@@ -0,0 +1,52 @@
+{ stdenv, fetchFromGitHub
+, meson, pkgconfig, ninja
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/default.nix:2]:
```suggestion
, meson, pkg-config, ninja
```

File: pkgs/applications/window-managers/wayfire/default.nix [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    owner = "WayfireWM";
+    repo = "wayfire";
+    rev = version;
+    sha256 = "sha256-r9k8e0LCDxr8tEZvcBbkvZ2RqcKXYX+/fxalKoqT15U=";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [ pkgconfig meson ninja ];
+  buildInputs = [
+    # egl glesv2
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/default.nix:27]:
```suggestion
```

File: pkgs/applications/window-managers/wayfire/default.nix [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    wayland wayland-protocols
+    cairo glm
+    libevdev freetype libinput
+    pixman libxkbcommon libdrm
+    libjpeg libpng
+    libGL mesa
+    libcap xcbutilerrors xcbutilwm libxml2
+  ];
+  mesonFlags = [
+    "-Duse_system_wlroots=disabled"
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/default.nix:37]: We want to use the system libs.

File: pkgs/applications/window-managers/wayfire/default.nix [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    "-Duse_system_wfconfig=disabled"
+    "-Dwlroots:logind-provider=systemd"
+    "-Dwlroots:libseat=disabled"
+  ];
+
+  enableParallelBuilding = true;
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/default.nix:43]:
```suggestion
```

Set by meson.

File: pkgs/applications/window-managers/wayfire/plugins.nix [hunk 9]
```diff
@@ -0,0 +1,55 @@
+{ stdenv, meson, ninja
+, fetchFromGitHub, wayfire, pkgconfig
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/plugins.nix:2]:
```suggestion
, fetchFromGitHub, wayfire, pkg-config
```

File: pkgs/applications/window-managers/wayfire/plugins.nix [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+stdenv.mkDerivation rec {
+  pname = "wayfire-plugins-extra";
+  version = "b2036fc4e5935bf459b1779b89b59a73be92da9f";
+
+  src = fetchFromGitHub {
+    owner = "WayfireWM";
+    repo = "wayfire-plugins-extra";
+    rev = version;
+    sha256 = "sha256-u1gOd6Sp2hDgOA+gln6C+ADyzrI0HrGQFu5L0zFpZuE=";
+    fetchSubmodules = true;
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/plugins.nix:19]: Please do not use the included dbus.

File: pkgs/applications/window-managers/wayfire/plugins.nix [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  };
+
+  nativeBuildInputs = [ pkgconfig meson ninja ];
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/plugins.nix:22]:
```suggestion
  nativeBuildInputs = [ pkg-config meson ninja ];
```

File: pkgs/applications/window-managers/wayfire/shell.nix [hunk 12]
```diff
@@ -0,0 +1,44 @@
+{ stdenv, fetchFromGitHub
+, meson, ninja, pkgconfig
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/shell.nix:2]:
```suggestion
, meson, ninja, pkg-config
```

File: pkgs/applications/window-managers/wayfire/shell.nix [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+  src = fetchFromGitHub {
+    owner = "WayfireWM";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-Ci66TOTyWZZ14VPJz3pvi+mI9EtCPMveDm5NS97l+Y4=";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [ meson ninja pkgconfig ];
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/shell.nix:21]:
```suggestion
  nativeBuildInputs = [ meson ninja pkg-config ];
```

File: pkgs/applications/window-managers/wayfire/plugins.nix [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  ];
+
+  mesonFlags = [
+    "-Denable_nk=true"
+    "-Denable_dbus=true"
+    "-Denable_windecor=true"
+  ];
+
+  enableParallelBuilding = true;
+
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/plugins.nix:47]:
```suggestion

```

File: pkgs/applications/window-managers/wayfire/shell.nix [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+stdenv.mkDerivation rec {
+  pname = "wf-shell";
+  version = "3fed0524897f297bc4e7c49049c6b4221a2ff417";
+
+  src = fetchFromGitHub {
+    owner = "WayfireWM";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-Ci66TOTyWZZ14VPJz3pvi+mI9EtCPMveDm5NS97l+Y4=";
+    fetchSubmodules = true;
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/shell.nix:18]:
```suggestion
```

File: pkgs/applications/window-managers/wayfire/shell.nix [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+  buildInputs = [
+    alsaLib glm gtkmm3
+    gtk-layer-shell
+    libpulseaudio
+    wayland librsvg
+    wayland-protocols
+    wayfire
+  ];
+
+  enableParallelBuilding = true;
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/shell.nix:31]:
```suggestion
```

File: pkgs/applications/window-managers/wayfire/shell.nix [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+
+  mesonFlags = [
+    "-Dpulse=enabled" "-Dwayland-logout=true"
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/shell.nix:34]:
```suggestion
    "-Dpulse=enabled"
    "-Dwayland-logout=true"
```

File: pkgs/applications/window-managers/wayfire/wcm.nix [hunk 18]
```diff
@@ -0,0 +1,36 @@
+{ stdenv, meson, ninja
+, fetchFromGitHub, wayfire
+, wf-shell, pkgconfig
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/wcm.nix:3]:
```suggestion
, wf-shell, pkg-config
```

File: pkgs/applications/window-managers/wayfire/wcm.nix [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+  version = "5d935e0727ab976c85690669083addc9c21ee298";
+
+  src = fetchFromGitHub {
+    owner = "WayfireWM";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-Nrz/Em80+C8uTUf0+Ny/WLcMKFVRKAzXaeIp3m7NsnU=";
+  };
+
+  nativeBuildInputs = [ meson ninja pkgconfig ];
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/wcm.nix:20]:
```suggestion
  nativeBuildInputs = [ meson ninja pkg-config ];
```

File: pkgs/applications/window-managers/wayfire/wcm.nix [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+  buildInputs = [
+    wayfire wf-shell
+    libevdev wayland
+    wayland-protocols
+    libxml2 gtk3 glm
+  ];
+
+
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/wayfire/wcm.nix:28]:
```suggestion
  ];

```

File: pkgs/top-level/all-packages.nix [hunk 21]
```diff
@@ -28752,4 +28752,10 @@ in
   psftools = callPackage ../os-specific/linux/psftools {};
 
   lc3tools = callPackage ../development/tools/lc3tools {};
+
+  wayfire = callPackage ../applications/window-managers/wayfire {};
+  wf-shell = callPackage ../applications/window-managers/wayfire/shell.nix {};
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:28757]: Please put those under window manager/libs and sort them alphabetically.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
Also please put each new package into its own commit and squash review changes together.

File: maintainers/maintainer-list.nix [hunk 22]
```diff
@@ -10106,4 +10106,9 @@
...
+    name = "Ishan Agarwal";
+    email = "ishan365.ia@gmail.com";
+    github = "ishan9299";
+    githubId = 47824004;
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:10113]:
```suggestion
    githubId = 47824004;
  };
```

File: pkgs/applications/window-managers/wayfire/default.nix (hunk 3 above)
Comment (ishan9299) [reply pkgs/applications/window-managers/wayfire/default.nix:22]: Yes we can in fact I was doing that before.

File: pkgs/applications/window-managers/wayfire/plugins.nix (hunk 10 above)
Comment (ishan9299) [reply pkgs/applications/window-managers/wayfire/plugins.nix:19]: This a plugin for dbus interface it is not dbus. I create a seperate package for it.

File: pkgs/applications/window-managers/wayfire/default.nix (hunk 3 above)
Comment (ishan9299) [reply pkgs/applications/window-managers/wayfire/default.nix:22]: There is a problem that wf-utils and wf-touch are expected to a subproject. The meson.build fails if we pass it as a dependency. We can do something like [this](https://github.com/NixOS/nixpkgs/blob/master/pkgs/desktops/gnome-3/core/gvc-with-ucm-prePatch.nix) to avoid pulling down submodules automatically.

Comment (SuperSandro2000) [reply pkgs/applications/window-managers/wayfire/default.nix:22]: I am not sure if this is the ideal solution. Do you know if upstream would support using system libraries? Otherwise I think this is fine if you do not know a better solution.

Comment (ishan9299) [reply pkgs/applications/window-managers/wayfire/default.nix:22]: Actually the I mentioned didn't work the build is failing. Also the submodules wayfire requires provide utility classes and support for touch. We could also try patching meson.build to add the wf-touch and wf-utils as dependencies and not as a submodule but I am not familiar with meson.build (I am just firing blind arrows after reading the meson.build file).

File: pkgs/applications/window-managers/wayfire/default.nix (hunk 7 above)
Comment (ishan9299) [reply pkgs/applications/window-managers/wayfire/default.nix:37]: Can't the system version and the version wayfire uses aren't compatible fails to build the packages.

File: pkgs/applications/window-managers/wayfire/default.nix (hunk 3 above)
Comment (ishan9299) [reply pkgs/applications/window-managers/wayfire/default.nix:22]: Ok after looking into the issue why the build was failing was due to wlroots in nixpkgs is incompatible with the version wayfire is using. So we will have to use the fetchSubmodules.

**Comment by ishan9299** [comment]
I am sticking with the submodules provided by the upstream as they don't seem to follow the latest or released version of the commits for their submodules.

**Comment by ishan9299** [comment]
Right the icons are missing in the menu and the wcm settings app.
```
(wcm:6724): Gtk-WARNING **: 19:37:24.796: Could not load a pixbuf from /org/gtk/libgtk/theme/Adwaita/assets/check-symbolic.svg.
This may indicate that pixbuf loaders or the mime database could not be found.

(wcm:6724): Gtk-WARNING **: 19:37:24.817: Could not find the icon 'window-minimize-symbolic-ltr'. The 'hicolor' theme
was not found either, perhaps you need to install it.
You can get a copy from:
	http://icon-theme.freedesktop.org/releases
```
Also the wayland-logout binary (submodule of wf-shell) complains about missing lsof command.
```
./result/bin/wayland-logout: line 24: lsof: command not found
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
```
I can't figure out how to fix these issues any help would be appreciated.

File: pkgs/applications/window-managers/wayfire/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/applications/window-managers/wayfire/default.nix:21]: Normally we try not to vendor packages that are already included in nixpkgs to reduce closure size.

**Comment by ishan9299** [comment]
I am going to redo the Pull Request. Sorry, I had a merge conflict and a file got removed so the whole thing is failing.
